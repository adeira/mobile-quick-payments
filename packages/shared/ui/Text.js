// @flow

import * as React from 'react';
import RN from 'react-native';
import { Translation } from 'mobile-quick-payments-translations';

import StyleSheet from './PlatformStyleSheet';
import Colors from './Colors';
import type { StylePropType } from '../index';

type Props = {|
  children?: React.Element<typeof Translation>,
  style?: StylePropType,
|};

export default class Text extends React.Component<Props> {
  // note: this must be class (not functional component) to work properly
  // with Animated library from RN

  render = () => (
    <RN.Text style={[styleSheet.title, this.props.style]}>
      {this.props.children ? React.Children.only(this.props.children) : null}
    </RN.Text>
  );
}

const styleSheet = StyleSheet.create({
  title: {
    fontSize: 16,
    fontWeight: 'normal',
    color: Colors.text,
  },
});
