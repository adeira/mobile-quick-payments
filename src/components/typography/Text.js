// @flow

import * as React from 'react';
import RN from 'react-native';
import Translation from '_translations';

import StyleSheet from '../stylesheet';
import Colors from '../Colors';
import type {StylePropType} from '../stylesheet';

type Props = {|
  +children?: React$Element<typeof Translation>,
  +style?: StylePropType,
|};

const TextContext: React$Context<{|
  +isNested: boolean,
|}> = React.createContext({
  isNested: false,
});

/**
 * Default styling must be applied only on the root Text component.
 */
const RootText = ({children, style}) => (
  <TextContext.Provider value={{isNested: true}}>
    <RN.Text style={[styleSheet.defaultText, style]}>{children}</RN.Text>
  </TextContext.Provider>
);

/**
 * Style property may be `undefined`. Style `undefined` in nested Text
 * components indicates style inheritance in RN.
 */
const NestedText = ({children, style}) => (
  <RN.Text style={style}>{children}</RN.Text>
);

export default class Text extends React.Component<Props> {
  // note: this must be class (not functional component) to work properly
  // with Animated library from RN

  render = () => {
    const {children} = this.props;

    return (
      <TextContext.Consumer>
        {({isNested}) => {
          return isNested === true ? (
            <NestedText style={this.props.style}>{children}</NestedText>
          ) : (
            <RootText style={this.props.style}>{children}</RootText>
          );
        }}
      </TextContext.Consumer>
    );
  };
}

const styleSheet = StyleSheet.create({
  defaultText: {
    fontSize: 16,
    fontWeight: 'normal',
    color: Colors.text,
  },
});
