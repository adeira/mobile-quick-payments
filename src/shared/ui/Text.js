// @flow

import * as React from 'react';
import { Text } from 'react-native'; // eslint-disable-line no-restricted-imports
import { Translation } from 'mobile-quick-payments-translations';

import StyleSheet from './PlatformStyleSheet';
import type { StylePropType } from '../index';

type Props = {|
  children: React.Element<typeof Translation>,
  style?: StylePropType,
|};

export default ({ children, style }: Props) => (
  <Text style={[styleSheet.title, style]}>{children}</Text>
);

const styleSheet = StyleSheet.create({
  title: {
    fontSize: 16,
    fontWeight: 'normal',
  },
});
