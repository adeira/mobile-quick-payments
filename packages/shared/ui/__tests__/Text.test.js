// @flow

import * as React from 'react';
import Renderer from 'react-test-renderer';

import Text from '../Text';

const VoidComponent = () => 'VOID';

it('is possible to use Text component with child', () => {
  expect(
    Renderer.create(
      <Text>
        {/* $FlowExpectedError: children should be type of Translation in real code */}
        <VoidComponent />
      </Text>,
    ),
  ).toMatchSnapshot();
});

it('is possible to use Text component without children', () => {
  expect(Renderer.create(<Text />)).toMatchSnapshot();
});

it('is possible to use Text component with custom styles', () => {
  expect(
    // $FlowExpectedError: children should be type of Translation in real code
    Renderer.create(
      <Text style={{ color: 'red' }}>
        {/* $FlowExpectedError: children should be type of Translation in real code */}
        <VoidComponent />
      </Text>,
    ),
  ).toMatchSnapshot();
});
