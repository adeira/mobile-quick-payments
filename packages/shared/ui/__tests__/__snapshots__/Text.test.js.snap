// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`is possible to use Text component with child 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "color": "#141414",
        "fontSize": 16,
        "fontWeight": "normal",
      },
      undefined,
    ]
  }
>
  VOID
</Text>
`;

exports[`is possible to use Text component with custom styles 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "color": "#141414",
        "fontSize": 16,
        "fontWeight": "normal",
      },
      Object {
        "color": "red",
      },
    ]
  }
>
  VOID
</Text>
`;

exports[`is possible to use Text component without children 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "color": "#141414",
        "fontSize": 16,
        "fontWeight": "normal",
      },
      undefined,
    ]
  }
/>
`;
